rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,416 @@
+- The stub selections are located adjacent to the harness, which makes it easy to understand which replacements are going to happen for each harness.
+
+### Risks
+
+- Users can always write stubs that do not correctly correspond to program behavior, and so a successful verification does not actually mean the program is bug-free.
+This is similar to other specification bugs. 
+
+### Comparison to function contracts
+
+- In many cases, stubs are more user-friendly than contracts. With contracts, it is necessary to explicitly provide information that is automatically captured in Rust (such as which memory is written).
+- The [currently proposed function contract mechanism](https://github.com/model-checking/kani/tree/features/function-contracts) does not provide a way to put contracts on external functions.
+This is one of the key motivations for stubbing.
+
+### Alternative #1: Annotate stubbed functions
+
+In this alternative, users add an attribute `#[kani::stub_by(<replacement>)]` to the function that should be replaced.
+This approach is similar to annotating a function with a contract specifying its behavior (the stub acts like a programmatic contract).
+The major downside with this approach is that it would not be possible to stub external code. We see this as a likely use case that needs to be supported: users will want to replace `std` library functions or functions from arbitrary external crates.
+
+### Alternative #2: Annotate stubs
+
+In this alternative, users add an attribute `#[kani::stub(<original>)]` to the stub function itself, saying which function it replaces:
+
+```rust
+#[cfg(kani)]
+#[kani::stub("rand::random")]
+fn mock_random<T: kani::Arbitrary>() -> T { ... }
+```
+
+The downside is that this stub must be uniformly applied across all harnesses and the stub specifications might be spread out across multiple files.
+It would also require an extra layer of indirection to use a function as a stub if the user does not have source code access to it.
+
+### Alternative #3: Annotate harnesses and stubs 
+
+This alternative combines the proposed solution and Alternative #2.
+Users annotate the stub (as in Alternative #2) and specify for each harness which stubs to use using an annotation `#[kani::use_stubs(<stub>+)]` placed above the harness.
+
+This could be combined with modules, so that a module can be used to group stubs together, and then harnesses could pull in all the stubs in the module:
+
+```rust
+#[cfg(kani)]
+mod my_stubs {
+
+  #[kani::stub("foo")]
+  fn stub1() { ... }
+
+  #[kani::stub("bar")]
+  fn stub2() { ... }
+
+}
+
+#(cfg[kani])
+#[kani::proof]
+#[kani::use_stubs("my_stubs")]
+fn my_harness() { ... }
+```
+
+The benefit is that stubs are specified per harness, and (using modules) it might be possible to group stubs together.
+The downside is that multiple annotations are required and the stub mappings themselves are remote from the harness.
```

Comment:
Conflicts wouldn't be easy to deal with, right? For example, if I had to stubs replacing the same function, I need to change one stub set so only one stub gets used for the replacement.

Reply:
Wouldn't this problem occur whenever you are using stub sets (e.g., also in the stub set approach we suggest)? Like our proposed approach, this alternative doesn't require using stub sets - it's just an option